### PR TITLE
Fix null encoding in new document in move type action

### LINKED ISSUE #72984: Hot Reload unusable with CS8055 if a file created via "Move type to X.cs" is not touched by the user
## Version Used
VS 2022 Version 17.9.5
Roslyn 4.9.0 (a98c90d5645537)

Project info
C# 12.0
.NET 8.0

## Steps to Reproduce

1. Create a file `XYZ.cs` that contains a type named XYZ
2. Within that same file declare another type named ABC (different than XYZ)
3. Use the "Move type to ABC.cs" code action
4. Run the program with a debugger
5. Pause execution at any point
6. Edit any file within the assembly containing the above types (XYZ and ABC)
7. Ensure that there are no compiler errors after the edits
8. Use Hot Reload to load the changes onto the program and continue execution

## Expected Behavior

Hot Reload works as expected, without any errors.

## Actual Behavior

Hot Reload fails, and the following is observed in the output of Hot Reload:
```
19:44 51.27 [Error] E:\repos\[truncated]\File.cs (line 1): error CS8055: Cannot emit debug information for a source text without encoding.
19:44 51.27 Invalid changes were found. Please refer to the Error List window to fix those issues.
```

The user is prompted to edit the code or rerun the program anew, after Hot Reload fails.

The bug can be worked around by manually changing and saving changes to the ABC file before hot reloading, without stopping the current execution. Once the file is manually changed, hot reloading works.

### PR DESCRIPTION
**Pull request by Rekkonnect**
Closes #72984

## Summary

We default to receiving the encoding set in the new node we are requested to create the new document with in `DocumentState`.

A new test was added to test how hot reload reacts by applying the code fix that contained the null encoding bug, and applying actual code changes after performing the move type. It seems that the bug was not being triggered if the code fix had no changes to apply, i.e. the code remained the exact same, and all that changed was the location of the moved type.

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]
seems reasonable.  @tmat @jasonmalinowski any concerns?

**Comment by CyrusNajmabadi** [comment]
@tmat ccan you ptal?

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 1]
```diff
@@ -499,8 +499,8 @@ internal DocumentState UpdateTree(SyntaxNode newRoot, PreservationMode mode)
         }
         else
         {
-            // the existing encoding was never observed so is unknown.
-            encoding = null;
+            // the existing encoding was never observed so we try to use the one from the desired root.
+            encoding = newRoot.SyntaxTree.Encoding;
```
[the remark this answers is not in the document]
Comment (Rekkonnect) [reply src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs:503]: So we don't want to preserve the previous encoding of the tree and override it with the new tree's encoding

[the remark this answers is not in the document]
Comment (Rekkonnect) [reply src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: Done

[the remark this answers is not in the document]
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: Why not? If the document is updated to contain new syntax tree it is odd that we ignore part of the state associated with that new tree (encoding).

**Comment by Rekkonnect** [comment]
@tmat could you please review this?

**Comment by Rekkonnect** [comment]
@tmat another reminder to review this

**Comment by Rekkonnect** [comment]
`MSBuildWorkspacePreservesEncoding` fails at `Assert.NotNull(doc3text.Encoding);` because of the last change where we always override the encoding of the syntax tree. This behavior must also be reflected in the test, so I'll modify that. We can also preserve the behavior of retaining the encoding if we already had an encoding and the new syntax tree has a `null` one.

**Comment by Rekkonnect** [comment]
@tmat ptal at the changed tests and review

**Comment by Rekkonnect** [comment]
@tmat ready for review

**Comment by Rekkonnect** [comment]
@tmat ready for review

**Comment by CyrusNajmabadi** [comment]
Set to auto-merge.  If we don't want this, please speak up @tmat

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ptal.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 2]
```diff
@@ -495,23 +495,8 @@ internal DocumentState UpdateTree(SyntaxNode newRoot, PreservationMode mode)
...
-        {
-            encoding = priorText.Encoding;
-        }
-        else
-        {
-            // the existing encoding was never observed so is unknown.
-            encoding = null;
-        }
+        // use the encoding that we get from the new root
+        var encoding = newRoot.SyntaxTree.Encoding;
```
[the remark this answers is not in the document]
Comment (Rekkonnect) [reply src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs:499]: Note the previous comment from tmat: https://github.com/dotnet/roslyn/pull/74623?new_mergebox=false#discussion_r1706248120

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs:499]: Ok. Seems reasonable. Merging in. If you have any concerns @jason let us know

**Pull request merged**